### ISSUE TITLE
Provide a Hero block feature out of the box.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1152]
Alright, I know that this can be achieved by contrib modules like [Nivo Slider](https://github.com/backdrop-contrib/nivo_slider) or [Views Slideshow](https://github.com/backdrop-contrib/views_slideshow), but I believe it is very common nowadays for people to want to design their website with either a single image section (banner) or a slider.

The single image (banner) section can be seen as the simplest version of the slider that only contains one slide. So, would it be too far-fetched to ask that this feature is offered by Backdrop out of the box?

This was kinda brought up in #1042 and I think would be a killer feature for Backdrop!! ...but because I see this easily ending in an endless debate over which slideshow module/library is "the best", I propose to base this on a view (already in core) of nodes that contain only image and description/link fields (soon in core) and make the slider library "plugable".

---

~~PR from @jenlampton https://github.com/backdrop/backdrop/issues/1523~~
~~PR from @quicksketch https://github.com/backdrop/backdrop/issues/1526~~
PR from @jenlampton https://github.com/backdrop/backdrop/pull/1528

**Comment by klonos** [comment]
...a few things to support my case:

[views_slideshow](https://www.drupal.org/project/views_slideshow) is used in 25% of D7 websites ([~250k installs of the 7.x version](https://www.drupal.org/project/usage/views_slideshow) when [D7 counts a bit over 1mil sites](https://www.drupal.org/project/usage/drupal)) ...that's one in every 4 websites has a slideshow. ...and that's only the ones using views_slideshow - there's other popular slider modules for D7 too. So, it is fair to assume that the banner/slider thing is not an uncommon feature.

I was taking a look at a couple of tutorials linked from the views_slideshow module page ([this one](http://www.brightwebsitedesign.com/how-to-install-views-slideshow-module-on-drupal7) and [this one](http://ryanszrama.com/blog/06-08-2010/building-rotating-image-banner-views-slideshow) in particular) and specifically their "required modules" sections. What you'll notice is that 90% of the required contrib modules are now in Backdrop core (or planned to be). So, why not?!?

**Comment by biolithic** [comment]
I'm not sure that front page slideshows actually are something besides
generic noise in a lot of cases or that I care about them on mobile.  I'm
ok with people making slideshow modules but I wouldn't want all Backdrop
sites to be forced to use them.  If someone wants to do this work and make
it optional, that's ok with me.  Also, education on the Nivo Slider because
it doesn't seem too hard to set up in a short amount of time.  Thanks

**Comment by MrHaroldA** [comment]
Well, I'm against sliders in general, but the statistics don't lie (this time).

I would, however, be a great addition to the "full install profile" if you'd get a fancy slider on your front page for free. Simpler install profiles should absolutely not include such a slider.

**Comment by klonos** [comment]
Yeah, I don't think that 25% of websites qualifies to have this feature on by default. All I'm saying is that instead of having 25% of the people (especially newcomers) go through tutorials is not ideal UX. Enabling a view provided by core, creating one or a few slide nodes (depending on whether a simple banner or a slider is required) and then placing a block in your front page layout is a lot easier compared to having them researching for, then downloading then enabling various contrib modules and then having to make them all "sing together" in harmony.

**Comment by docwilmot** [comment]
> the only slider module ported to Backdrop at the time of writing this AFAIK

[Views Slideshow](https://github.com/backdrop-contrib/views_slideshow) already ported by @Graham-72.

I think though that this probably works better as contrib. Keeping in mind that once we've got Project Browser in core, discovering and installing new functionality will be much easier. Users can then freely select their own choice, avoiding the `endless debate over which slideshow module/library is "the best"`. Features like this I feel can live outside core to ease the maintenance burden.

**Comment by klonos** [comment]
> [Views Slideshow](https://github.com/backdrop-contrib/views_slideshow) already ported by @Graham-72.

Sorry, I missed that. Thanx.

PS: updating the issue summary to include it...

**Comment by klonos** [comment]
> Keeping in mind that once we've got Project Browser in core, discovering and installing new functionality will be much easier.

Looks like you're right on this one. Shoot the issue down then?

**Comment by quicksketch** [comment]
> Looks like you're right on this one. Shoot the issue down then?

I still like the idea of having some kind of big feature item out-of-box (as suggested in #1042), but this may not be the right solution.

Other (kind of crazy) ideas: Does this need to be Views based (and pluggable)? What if we left the more complex use-case to contrib and instead made a module that specifically provided a basic slider? Something like:

- A module that implements hook_block() to make sliders.
- An interface to edit sliders.
- Each slider had a multiple value field that supported 4 things: image, description, action text, action URL.
- All sliders are stored as config, rather than content.
- Ship with a popular library in core for sliders (I'd probably say https://github.com/woothemes/FlexSlider)

OTOH, maybe this would just make another contrib module. I just wanted to turn all the requirements around and see how it felt.

And maybe a slider overall is the wrong idea for core. But the popularity is undeniable, and WordPress does include an option out-of-box (as a "quicktag" aka "token"). Not pluggable either, but Backdrop isn't WordPress.

**Comment by docwilmot** [comment]
First hit on a Google search for "website sliders": [Why Sliders Make Your Website Suck](https://thrivethemes.com/why-sliders-suck/)

If we think they are wrong, then ignore them and read on:

I had been considering retooling my port of Nivo Slider to use the new UI suggested for the filters page - ie draggable rows, with dialog for configuration. I don't really like the vertical tabs which it uses now.

So it would be draggable rows with a thumbnail and a trim of the slide text, etc. And popup dialog to configure text, link and slide image, or to add a new slide. If someone can say if (apart from the UI) the current code in Nivo is reasonably Kosher for core, I can adapt it to use Flexslider instead and we could add this to core.

**Comment by quicksketch** [comment]
> Each slider had a multiple value field that supported 4 things: image, description, action text, action URL.
> All sliders are stored as config, rather than content.

BTW, after thinking on this for a few days, I decided this is sort of a terrible idea. Because sliders are likely to be content, if we provided such a tool, we may end up with sites that are just made of hundreds of configured sliders. All of which would be managed in config (incorrectly), instead of being made up of content. Encouraging users to assemble large swaths of content that is all stored in config would be a dangerous precedent, as configuration items are not accessible via Views and not likely to be very scalable (e.g. thousands of slider items).

Instead, I think what may be a good approach for the *content* of sliders would be to continue to improve our File entity handling. The thing that drives me nuts about sliders based on nodes is that you end up with slider items that shouldn't have full URLs being accessible individually. Then you need something like Rabbit Hole module to suppress their URLs. If files progressed to be fully fieldable entities (which is the plan anyway), then you could at categories for sliders or make a file type specifically that was for slider items.

**Comment by jenlampton** [comment]
> Instead, I think what may be a good approach for the content of sliders would be to continue to improve our File entity handling. 

I'm leaning this way too. If we were able to solve our file entity problems we may also be able to fix https://github.com/backdrop/backdrop-issues/issues/1373 rather than just "turning off" the feature which is broken.

I'm envisioning this: 
* a custom block with a multi-value file field.
* files can be uploaded directly through the block UI).
* what's saved in config is the file IDs, not the files themselves. 
* the block is rendered using a super lightweight CSS-only responsive slider, something like this: http://qnimate.com/creating-a-slider-using-html-and-css-only/#prettyPhoto

**Comment by jenlampton** [comment]
Here's what we thought would be possible for 1.5:
* Provide a new `Hero` block type in core. Identical to the custom block, but also has an image field.
* Without an image, the theme's CSS can place a background-image on the hero.
* With an image, it gets inline'd on the wrapper for the block, and the UI-provided image wins.

I'm working on a PR and should have something ready soon to test!

**Comment by jenlampton** [comment]
I just pushed a PR to https://github.com/backdrop/backdrop/issues/1523 for feature-review. I need to test the custom/stand-alone block functionality, and if we like this the PR will need tests. I'd love to get some eyes on it now though, let me know what you think?

**Comment by olafgrabienski** [comment]
Nice start!

**First thoughts**

- Let's add the ability to set the font size of the block content. It's quite small at the moment and doesn't look very 'heroic'. As as site builder / themer I can do it by myself via the theme css, but I guess the hero block type is primarily for other users.
- Should we provide instructions / help text about the recommended size of the background image?

**Some observations**

- If the text is 'longer' as the image height, the image gets repeated.
- I saw this error message while editing the hero block:
```
Notice: Undefined index: label in BlockText->formValidate() (line 115 of /home/qa/github/backdrop/backdrop/pr/1523/core/modules/layout/includes/block.text.inc).
Notice: Undefined index: delta in BlockText->formValidate() (line 116 of /home/qa/github/backdrop/backdrop/pr/1523/core/modules/layout/includes/block.text.inc).
Notice: Undefined index: reusable in BlockText->formValidate() (line 117 of /home/qa/github/backdrop/backdrop/pr/1523/core/modules/layout/includes/block.text.inc).
Notice: Undefined index: reusable in BlockText->formSubmit() (line 134 of /home/qa/github/backdrop/backdrop/pr/1523/core/modules/layout/includes/block.text.inc).
```

**Comment by docwilmot** [comment]
Agreeing with @olafgrabienski re controls for font size. I think even a couple of header types fields built in (hero-title, hero-subtitle, hero-body, hero-footer). 

Would also be great to match the display height to the image height. 

This would of course make this fundamentally different from a plain text block. As commented on the PR, we could probably consider extending `Block` and make this a different thing from text blocks. Not sure how likely you will need to reuse a hero block.

**Comment by jenlampton** [comment]
>Let's add the ability to set the font size of the block content

This isn't a pattern we do anywhere else in core, and I think it's a bad idea to start now. font-size isn't semantic, and shouldn't be controlled via a UI.

> It's quite small at the moment and doesn't look very 'heroic'.

This problem could be solved by adding some default styles for hero blocks into core themes. Bartik and Basis, for example, could double or triple hero block font size.

> As as site builder / themer I can do it by myself via the theme css

You also have two other UIs that you can use to control the semantics of what's going on. 1) use the dynamic block template to change your HTML elements. 2) Use the rich-text editor to mark up the content in the block however you want.

> Should we provide instructions / help text about the recommended size of the background image?

Yes, I think that would be a nice touch. Especially since the site-builder might be only looking at this on a desktop and forget that this also has to work on tiny screens.

> If the text is 'longer' as the image height, the image gets repeated.

ah, a little n`no-repeat` will fix that!

> I saw this error message 

def a bug, looking at it now! ty :)

> I think even a couple of header types fields.

We have another separate issue for hero/sliders, fieldable-blocks, or page-less-nodes, that will solve that problem shortly. Adding more fields here gets really quickly into the slippery-slope that is that issue. Baby steps, let's solve the most common use-case (single background image) first. The rich-text body of the block will serve as a suitable alternative for fields for the moment.

> Would also be great to match the display height to the image height.

I'm not sure it's possible to make generic. Which size device are you talking about? (see the problem?) Again this is a job for the theme. We can add a height in Basis, which will add a background image (and we'll know it's height). But to do it generically ... gets trixy.

> Not sure how likely you will need to reuse a hero block.

In practice, It's actually not that uncommon. Think of a hero block that promotes a thing. You might want that both on the home page and on the thing page, and maybe also on the thing's parent's page, thing's category page, etc. Anyway, I also agree it's far less important, but would be nice to have the option in the future.

**Comment by docwilmot** [comment]
>We have another separate issue for hero/sliders

Then why not just make a background option standard for text blocks instead of a special text block where the only difference is a background?

> I'm not sure it's possible to make generic.

Its likely possible with JS, not sure with CSS. Reason I recommend that is your background is either going to be longer than the enclosing DIV, and be chopped off or require endless line breaks to enlarge the div (yuck) or too short and end up showing white space below. You'll never have an image that is just the right height for any text you want in there. The slider type things (Nivo etc) calculate the height with JS to prevent the above.

**Comment by jenlampton** [comment]
> Then why not just make a background option standard for text blocks

That is an option, I suppose. From the user experience side of things I'd definitely not think about adding  text block though when all I wanted was a big image (and no text). I think having something with a different name will make more sense to people, though we could have just added a separate `block_info` item that pointed at the same handler. I wonder how much confusion that would cause... 

I do also like that we have the ability to style all hero blocks separately from text blocks. Separate template file, separate CSS identifier. Both architecturally and conceptually they feel like different things. As do 'content blocks'. *shrug* 

> You'll never have an image that is just the right height for any text you want in there. 

And the text will change `height` on every device, too! Text will get longer, image will get shorter, very trixy indeed. Any ideas for a good solution? Maybe we just remove the background image for smaller screen sizes?

**Comment by jenlampton** [comment]
PR updated with fixes recommended above. 

I do still need some ideas for the help text, which will follow from what we decide to do with the image size issue.  I just read a few blog posts on the subject and some people are recommending setting a specific aspect ratio, 16:9, for example. They then retain the same aspect ratio on all screen-sizes all the way down. How would people feel about a fixed ratio?  It certainly would make the help text easier! :)

**Comment by docwilmot** [comment]
I would support a fixed aspect ratio; that'll make things more sane. We can then just force overflow:hidden on the text. 

16:9 might still be tall for a full width hero on wide screens though, no? 21:9? Maybe even option to select aspect ratio? Too much? :fearful:

**Comment by jenlampton** [comment]
> Maybe even option to select aspect ratio? Too much? 

too much :) Let me try 21:9 and see what that looks like.

**Comment by olafgrabienski** [comment]
Providing some feedback after re-testing:

**Height,** as long as we have no JS solution, we could try `background-size: cover;` as a first workaround. If the text isn't too long, the image should be the same size as without the definition. If the text is too long, the image detail will be changed, but it covers the size of the text content. I guess, that's a little better than white text below the image.

**More styling?** Should we for instance add some margins to the hero block? Example: Without margin-bottom the hero looks strange in the content region, if there is no other content on the page, due to missing margin to the footer. There will be some similar cases depending on theme, layout and region. I guess, it'll be better to do that in themes like Bartik and Basis ...

**Do some of the formatting via CKeditor:** Good point, but for that purpose I miss (1) the _Format_ drop down and (2) the link button in the editor.

**Btw, links**: Link text is a light blue, often it won't be light enough. Should we add a special style for links, e.g. a flat-design-button one?

**Help text:** I'll try to help but let's do that when the thing is finished.

**Comment by jenlampton** [comment]
> Should we for instance add some margins to the hero block?

Yes, this is a good idea. good thinking!

> I miss (1) the Format drop down and (2) the link button in the editor.

Then you can add them via editor config. (If there is a bug there, that is a separate issue, but likely also needs to be addressed! yes, OMG WHERE IS THE LINK BUTTON?!?)

> Link text is a light blue.

Hm, this is definitely a bartik/basis issue.

**Comment by klonos** [comment]
>>It's quite small at the moment and doesn't look very 'heroic'.

> This problem could be solved by adding some default styles for hero blocks into core themes. Bartik and Basis, for example, could double or triple hero block font size.

Sounds reasonable too me 👍 ...and pretty much everything else discussed above too. I agree: baby steps. As people start using the feature, we'll get more use cases that need to be covered and solutions will emerge (either adapting core or solving in contrib or the theme layer).

**Comment by jenlampton** [comment]
I updated the PR. Stream of thought: I set the aspect ratio on the hero block  to 21:9 (much better size @docwilmot thank you). I put the styles for that in Bartik: both with CSS, and a new template for hero blocks.

If we add add the help text onto the Hero block image field, I wonder if the hero block styling should live in core `system.theme.css` instead?  I wouldn't want people to switch themes and then have that help text not match the image aspect ratio anymore.

Ideas?

**Comment by olafgrabienski** [comment]
> I wonder if the hero block styling should live in core system.theme.css instead?

Makes sense from the user's perspective. If it's okay from the core developer perspective as well, let's do it.

**Comment by quicksketch** [comment]
Wow, that screenshot really drives home how much we should definitely remove the Title from the home page layout in #1042.

PR is looking really good but I left some feedback: https://github.com/backdrop/backdrop/pull/1523

I think we may be able to get some additional simplification yet.

**Comment by jenlampton** [comment]
@quicksketch can you address some of the questions asked in https://github.com/backdrop/backdrop-issues/issues/1152#issuecomment-245434427? 

I think you also had some ideas about how to solve the **height problem** that didn't involve an exact aspect ratio and allowed for a smidge more flexibility in images uploaded. I don't remember what it was though, could you please write it here?

**Comment by quicksketch** [comment]
I was thinking we might use media queries to set a minimum height to maintain the "hero banner" appearance, rather than basing the size of the aspect ratio of the image. So on a hero with only a little content, the size would jump in between breakpoints, roughly keeping a similar aspect ratio but not holding to it strictly.

After looking at other heros, it looks like they regularly have a problem when the content fits on larger screens but doesn't on smaller ones. We can't just keep shrinking the text, because the hero looses its purpose and legibility when the font shrinks proportionally with the image.

So as such, I think we should put less emphasis on the ratio of the uploaded image. Although a landscape picture will definitely look better in most cases, by the time you get to mobile size a hero very well may be portrait-shaped if the content is allowed to push the height of the hero.

Using `background-size: cover` instead of `background-size: 100%` means that the image will scale so that it always covers the entire container, even if it means cropping either from the top/bottom or sides (you can choose which sides get cropped by using `background-position`). So with this approach, you get something like this:

![peek 2016-09-08 22-12](https://cloud.githubusercontent.com/assets/100206/18376159/583708d0-7611-11e6-8e27-8e08a0d7ee91.gif)

Now that acts a lot like a 100% wide (aspect-ratio) block would, but with jumps in between sizes.

But look what happens with a lot more content (pardon the gif quality):

![peek 2016-09-08 22-11-2](https://cloud.githubusercontent.com/assets/100206/18376223/f5bfb1b0-7611-11e6-8409-b29329ddb54b.gif)

Note that at the smallest size, the hero is actually a portrait image, with the left/right cropped off. Font size is adjusted down slight on smaller screens but maintains its larger size than normal content, and certainly doesn't scale down just to fit with a landscape-size ratio image.

So this compromise results in a few things:

- No new block template or markup needed.
- Media queries used to set "minimum" (but not maximum") height.
- Aspect ratio is not guaranteed in the size of the hero. It will get taller on narrower screens.
- But the image is never stretched or distorted, it's cropped off on the left/right/bottom as needed (with background-position: top center, the top will never be cropped as shown here).

I'll push a PR with this demonstrated in code in a minute.

**Comment by quicksketch** [comment]
PR with the above approach (building on @jenlampton's PR): https://github.com/backdrop/backdrop/pull/1526

**Comment by quicksketch** [comment]
Note I don't think the "reusable" checkbox works with this yet. We could disable this ability for hero blocks initially and possibly add it later by disabling the reusable options:

```
    parent::form($form, $form_state);
    $form['convert']['#access'] = FALSE;
``

**Comment by jenlampton** [comment]
> Note I don't think the "reusable" checkbox works with this yet. 

I had them disabled in my PR with #access = FALSE.

**Comment by jenlampton** [comment]
@quicksketch in the future can you link your PR from the first post in the thread?  (see above)

**Comment by jenlampton** [comment]
Reactions:
* I love the `background-cover` affect, thank you @olafgrabienski for the suggestion! 
* Love that there is no new block template or markup needed with this approach :)
* I don't really like the snapping that happens to the hero for different screen sizes due to the min-heights added. I prefer how it responds without them (and I prefer that it's less CSS to override).  That said, I'd be fine with setting min-heights in Bartik if others agree. 
* I think forcing uploaded images to be a 2:1 ratio isn't going to be flexible enough for our users. (I personally like mine at more of a 4:1 ratio). If we left the "maximum" at 2:1 but allowed the minimum to be 4:1 then I think everyone wins. No?
* I like the addition of validators, but I don't like what it does to our help text. at this point I'd prefer not to add anymore description text, because the more we add the less likely people are to read it.
* The `hero-block` class is no longer properly added, tiny bug.

I'm pushing a PR with these three changes (lower min size, removed help text, `hero-block` class bug fixed) but after that this is RTBC from me.

**Comment by quicksketch** [comment]
> I like the addition of validators, but I don't like what it does to our help text. at this point I'd prefer not to add anymore description text, because the more we add the less likely people are to read it. I'm also not sure the text we added adds any value. Just more words, so...

If you mean we remove "The uploaded image will be displayed as a background" but leave the rest, then I agree. I think the allowed file types and size information is quite valuable and I wouldn't want to remove it.

Marking this needs work as the tests aren't passing yet, and we need to remove the "Resusable" checkbox for this block or get it working before merging.

**Comment by jenlampton** [comment]
> If you mean we remove "The uploaded image will be displayed as a background" but leave the rest

Yep :)

okay, I'll go back to those tests tonight. I still need to figure out how to upload an image in tests!

**Comment by quicksketch** [comment]
@wesruv left a lot of great feedback in the PR https://github.com/backdrop/backdrop/pull/1528

> I don't really like the snapping that happens to the hero for different screen sizes due to the min-heights added. I prefer how it responds without them (and I prefer that it's less CSS to override). That said, I'd be fine with setting min-heights in Bartik if others agree. 

To which Wes responded:

> I did see @jenlampton mentioned that the 'clunkiness' of the breakpoints wasn't great, we could add:
```
  -webkit-transition: min-height 0.25s;
  transition: min-height 0.25s;
  will-change: min-height;
```

Which sounds great. Although entirely eye-candy, it will make for smoother appearance when demoing any responsiveness of the homepage, which shouldn't be underestimated.

**Comment by jenlampton** [comment]
jeeeeeeeeeeez those tests were rough... I ended up brut-forcing the last check there, but I got them all passing. The tests even caught a bug in the process :) Anywho I pushed a new PR and these should - hopefully - go green. And we need more more final review!

Changes include:
* class name is now `block-hero` instead of `hero-block`
* added utility classes `block-hero-image` and `block-hero-no-image`
* image-only blocks now appear properly.
* images can now be removed from blocks.
* fancy css transitions courtesy of @wesruv 
* fancy background-cover courtesy of @olafgrabienski 
* HAS NEW TESTS that should pass
* moved block tests into separate test class, and removed some `t()`s from them

I think that's it...

**Comment by quicksketch** [comment]
So this is needs review then? The test regex is definitely wonky, but I would be okay with merging and trying to address that later (if at all). The test confirms what it needs to, although it's definitely odd whatever is happening in SimpleTest/SimpleXML that is making that so difficult to assert a `&nbsp;` string.

**Comment by jenlampton** [comment]
yes sorry, back to needs review.

**Comment by quicksketch** [comment]
https://github.com/backdrop/backdrop/pull/1528 looks good to me with minor fixes to the tests. I removed the "Reusable" checkbox until we can get that fully supported. Waiting on final test run for merge.

**Comment by jenlampton** [comment]
> I removed the "Reusable" checkbox until we can get that fully supported.

So did I! We should make sure that my #access = FALSE wasn't added to the wrong place...

**Comment by quicksketch** [comment]
Ah, sure enough. Sorry I didn't see that feedback addressed in the PR because you put the line at the bottom of the function instead of the top where the line comment was. I'll remove mine in the fix-up commit.

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/1528 for 1.5.0. Great job owning this one @jenlampton! Thanks everyone for the valuable feedback and reaching a compromise on this very slippery slope of functionality.

**Comment by klonos** [comment]
I don't have much time to test, but at least I can contribute by doing trivial things like creating follow-ups that we shouldn't forget. Hope it will save you guys a few seconds of your life 😉

**Comment by quicksketch** [comment]
I made a few additional follow-ups:

- [Change big 'ol "Welcome to Backdrop CMS" to "Welcome to [sitename]"](https://github.com/backdrop/backdrop-issues/issues/2165)
- [Change the $delta of the Hero block from "hero_block" to just "hero"](https://github.com/backdrop/backdrop-issues/issues/2167)
